<img src="https://drive.google.com/uc?export=view&id=1QyjJW7NL2WDfnVZ0_MqAfH6Slkal1DJA" width="150">

$\tiny{Universidad \ Nacional \ de \ San \ Agustín \ de \ Arequipa \\
Ciencia \ de \ la \ Computación \ UNSA}$

---

# <center>  $Capítulo \ 5$
### <center> $Cadenas \ de \ Markov$

**Sesión de Laboratorio 13 :** Cadenas de Markov

**Profesor:** ......................................

**Estudiante:** .........................................

---


# **Cadenas de Markov**

Una cadena de Markov es una sucesión de ensayos similares u observaciones en la cual cada ensayo tiene el mismo número finito de resultados posibles y en donde también la probabilidad de cada resultado para un ensayo dado depende sólo del resultado del ensayo inmediatamente precedente y no de cualquier resultado previo

Una cadena de Markov es un proceso estocástico, para el cual la sucesión  $\left\{X_t / t \geq 0\right \}$ que describe dicho proceso debe verificar la siguiente condición

$$P[X_{t+1}=j/ X_0=i_0,X_1=i_1,,...,X_{t-1}=i_{t-1},X_t=i]=P[X_{t+1}=j/ X_t=i]$$

**Matriz de estado**

$$p(k)=[p_1(k) \hspace{0.5cm} p_2(k) \hspace{0.5cm}...\hspace{0.5cm}p_m(k)]$$

$$p_1(k) + p_2(k) +...+p_m(k)=1$$

**Matriz de transición**


$$P=
\left[\begin{array}{lcr}
p_{11} & p_{12} & \dots & p_{1m}\\
p_{21} & p_{22} & \dots & p_{2m}\\
\vdots & \vdots & \ddots & \vdots\\
p_{m1} & p_{m2} & \dots & p_{mm}\\
\end{array}\right]
$$

$0 \leq p_{ij}\leq 1$

$$p_{i1} + p_{i2} + \dots + p_{im}=1, \forall i=1,2,...,m$$

**Teorema:** Si $P$ denota la matriz de transición de una cadena de Markov  y $p(k)$ es la matriz de estado después de $k$ ensayos entonces la matriz de estado $p(k+1)$ después del ensayo siguiente esta dada por $p(k+1)=p(k)P$

usando el teorema se concluye que:

$$p(k)=p(0)P^k$$

**Matriz de estado estacionaria**

$$\pi=[\pi_1 \hspace{0.5cm} \pi_2 \hspace{0.5cm}...\hspace{0.5cm}\pi_m], \pi_1 + \pi_2 +...+\pi_m=1$$

$$\pi_1 + \pi_2 +...+\pi_m=1$$

Para cualquier matriz de estado $X$, la matriz estacionaria verifica

$$\lim_{k \to \infty}XP^k=\pi$$

la matriz estacionaria se halla resolviendo el sistema

$$
\left\{
\begin{array}{ll}
\pi=\pi P\\
\sum \pi_i=1
\end{array}
\right.
$$




---

**Ejemplo:** Un sistema de transporte masivo entra en operación. Las autoridades de transito han realizado estudios que predicen el porcentaje de quienes utilizaran el sistema masivo(M) y el de los que seguirán utilizando su auto (A). Se ha obtenido la siguiente matriz de transición

<center>$El \ próximo \ año$
    
|$ $|$ $|$M$|$A$|
|- |--|--|--|
|$ $|$M$|0.7|0.3|
|$Este \ año$|$A$|0.2|0.8|


Suponga que la población del área permanece constante y que al principio el 30% de la gente se traslada en el transporte masivo y el 70% en auto.

¿Qué porcentaje utilizará el sistema de transporte masivo después de un año? ¿Después de dos años?¿Despues de un tiempo prolongado?

**Solución**

$p(0)=[0.3 \ 0.7]$



$p(1)=?$

$p(2)=?$

<center>
$P=
\left[\begin{array}{lcr}
0.7 & 0.3\\
0.2 & 0.8
\end{array}\right]
$

$p(1)=p(0)P$

In [ ]:
p0=c(0.3,0.7)
p0

[1] 0.3 0.7

In [ ]:
P=matrix(c(0.7,0.2,0.3,0.8),nrow=2,ncol=2)
print(P)

     [,1] [,2]
[1,]  0.7  0.3
[2,]  0.2  0.8


In [ ]:
p1=p0%*%P
print(p1)

     [,1] [,2]
[1,] 0.35 0.65


Luego de un año el 35% de las personas utilizará el transporte masivo y 65% usará su auto

$p(2)=p(1)P$

In [ ]:
p2=p1%*%P
print(p2)º

      [,1]  [,2]
[1,] 0.375 0.625


Luego de dos años el 37.5% de las personas utilizará el transporte masivo y 62.5% usará su auto

Equivalentemente

$p(2)=p(0)P^2$

## **Potenciación de matrices**

Para hallar la potencia de una matriz $A^n$
se puede utilizar la instrucción **A\% \^\%k**

para usar esta instrucción debemos utlizar el paquete **expm**

In [ ]:
install.packages("expm")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
library(expm)

Loading required package: Matrix


Attaching package: ‘Matrix’


The following objects are masked from ‘package:pracma’:

    expm, lu, tril, triu



Attaching package: ‘expm’


The following object is masked from ‘package:Matrix’:

    expm


The following objects are masked from ‘package:pracma’:

    expm, logm, sqrtm




In [ ]:
p2=p0%*%(P%^%2)
p2

[,1]  [,2] 
[1,] 0.375 0.625

## **Potenciación de matrices**

Para hallar la potencia de una matriz $A^n$
se puede utilizar la instrucción $mtx.exp(A,n)$

para usar esta instrucción debemos utlizar el paquete $Biodem$

In [ ]:
install.packages("Biodem")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
library(Biodem)

In [ ]:
mtx.exp(P,2)

[,1] [,2]
[1,] 0.55 0.45
[2,] 0.30 0.70

In [ ]:
p2=p0%*%mtx.exp(P,2)
print(p2)

      [,1]  [,2]
[1,] 0.375 0.625


Si calculamos diferentes matrices ce transicón $p(k)$ cuando $k$  tiende al infinito tenemos:


In [ ]:
p3=p0%*%mtx.exp(P,3)
print(p3)

       [,1]   [,2]
[1,] 0.3875 0.6125


In [ ]:
p4=p0%*%mtx.exp(P,4)
print(p4)

        [,1]    [,2]
[1,] 0.39375 0.60625


In [ ]:
p5=p0%*%mtx.exp(P,5)
print(p5)

         [,1]     [,2]
[1,] 0.396875 0.603125


In [ ]:
p15=p0%*%mtx.exp(P,15)
print(p15)

          [,1]      [,2]
[1,] 0.3999969 0.6000031


In [ ]:
p50=p0%*%mtx.exp(P,50)
print(p50)

     [,1] [,2]
[1,]  0.4  0.6


Calculamos ahora diferentes potencias de la matriz de transición

$$P^k$$

In [ ]:
P2=mtx.exp(P,2)
print(P2)

     [,1] [,2]
[1,] 0.55 0.45
[2,] 0.30 0.70


In [ ]:
P3=mtx.exp(P,3)
print(P3)

      [,1]  [,2]
[1,] 0.475 0.525
[2,] 0.350 0.650


In [ ]:
P4=mtx.exp(P,4)
print(P4)

       [,1]   [,2]
[1,] 0.4375 0.5625
[2,] 0.3750 0.6250


In [ ]:
P5=mtx.exp(P,5)
print(P5)

        [,1]    [,2]
[1,] 0.41875 0.58125
[2,] 0.38750 0.61250


In [ ]:
P6=mtx.exp(P,6)
print(P6)

         [,1]     [,2]
[1,] 0.409375 0.590625
[2,] 0.393750 0.606250


In [ ]:
P7=mtx.exp(P,7)
print(P7)

          [,1]      [,2]
[1,] 0.4046875 0.5953125
[2,] 0.3968750 0.6031250


In [ ]:
P10=mtx.exp(P,10)
print(P10)

          [,1]      [,2]
[1,] 0.4005859 0.5994141
[2,] 0.3996094 0.6003906


In [ ]:
P20=mtx.exp(P,20)
print(P20)

          [,1]      [,2]
[1,] 0.4000006 0.5999994
[2,] 0.3999996 0.6000004


In [ ]:
P30=mtx.exp(P,30)
print(P30)

     [,1] [,2]
[1,]  0.4  0.6
[2,]  0.4  0.6


Intuitivamente tenemos

$$\lim_{k \to \infty}P^k=P_0$$


donde


<center>
$P_0=
\left[\begin{array}{lcr}
0.4 & 0.6\\
0.4 & 0.6
\end{array}\right]
$

## **Cálculo de la matriz estacionaria**

Es posible usar la libreria $Biodem$ o la libreria $pracma$

In [ ]:
install.packages("pracma")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
library("pracma")

## **Contrucción de la matriz aumentada**

In [ ]:
P=matrix(c(0.7,0.2,0.3,0.8),nrow=2,ncol=2)
print(P)
P2=P

     [,1] [,2]
[1,]  0.7  0.3
[2,]  0.2  0.8


In [ ]:
I=diag(1,2,2)
print(I)

     [,1] [,2]
[1,]    1    0
[2,]    0    1


In [ ]:
PP=t(P2)-I
PP

[,1] [,2]
[1,] -0.3  0.2
[2,]  0.3 -0.2

In [ ]:
orden=dim(P2)
orden

[1] 2 2

In [ ]:
b=matrix(0,orden[1],1)
b

[,1]
[1,] 0   
[2,] 0

In [ ]:
T=cbind(PP,b)
T

[,1] [,2] [,3]
[1,] -0.3  0.2 0   
[2,]  0.3 -0.2 0

In [ ]:
Aum=rbind(T,matrix(1,1,orden[1]+1))
Aum

[,1] [,2] [,3]
[1,] -0.3  0.2 0   
[2,]  0.3 -0.2 0   
[3,]  1.0  1.0 1

**Solución del sistema**

In [ ]:
rref(Aum)

[,1] [,2] [,3]
[1,] 1    0    0.4 
[2,] 0    1    0.6 
[3,] 0    0    0.0

## **Construcción y solución del sistema**

In [ ]:
P=matrix(c(0.7,0.2,0.3,0.8),nrow=2,ncol=2)
print(P)
P2=P
orden=dim(P2)
I=diag(1,orden)
PP=t(P2)-I
b=matrix(0,orden[1],1)
T=cbind(PP,b)
Aum=rbind(T,matrix(1,1,orden[1]+1))
B=rref(Aum)
B2=B[1:orden[1],orden[1]+1]
cat("Matriz estacionaria\n")
print(B2)

     [,1] [,2]
[1,]  0.7  0.3
[2,]  0.2  0.8
Matriz estacionaria
[1] 0.4 0.6


Hoy día tres empresas procesadoras de productos X,Y,Z controlan el 50%, 30% y 20% del mercado de café respectivamente. Al mismo tiempo las tres empresas presentan
marcas de café con la introducción de estas nuevas marcas en un año ocurre lo siguiente:

a)X retiene el 60% de sus consumidores y cede el 20% a Y y otro 20% a Z

b)Y conserva el 50% de sus consumidores y pierde el 30% con X y un 20% con Z

c)Z retiene el 70% de sus consumidores, cede el 10% a X y el 20% a Y

Suponiendo que esta tendencia continua que proporción del mercado tendrá cada
empresa al termino de dos años ¿Qué proporción del mercado tendrá cada empresa a
largo plazo?

In [ ]:
p0=c(0.5,0.3,0.2)
P=matrix(c(0.6,0.3,0.1,0.2,0.5,0.2,0.2,0.2,0.7),nrow=3,ncol=3)
p1=p0%*%P
p1


[,1] [,2] [,3]
[1,] 0.41 0.29 0.3

In [ ]:
p2=p1%*%P
p2

[,1]  [,2]  [,3]
[1,] 0.363 0.287 0.35

In [ ]:
P%^%2

[,1] [,2] [,3]
[1,] 0.44 0.26 0.30
[2,] 0.35 0.35 0.30
[3,] 0.19 0.26 0.55

In [ ]:
p2=p0%*%(P%^%2)
p2

[,1]  [,2]  [,3]
[1,] 0.363 0.287 0.35

In [ ]:
P=matrix(c(0.6,0.3,0.1,0.2,0.5,0.2,0.2,0.2,0.7),nrow=3,ncol=3)
print(P)
P2=P
orden=dim(P2)
I=diag(1,orden)
PP=t(P2)-I
b=matrix(0,orden[1],1)
T=cbind(PP,b)
T
Aum=rbind(T,matrix(1,1,orden[1]+1))
B=rref(Aum)
B
B2=B[1:orden[1],orden[1]+1]
cat("Matriz estacionaria\n")
print(B2)

     [,1] [,2] [,3]
[1,]  0.6  0.2  0.2
[2,]  0.3  0.5  0.2
[3,]  0.1  0.2  0.7


[,1] [,2] [,3] [,4]
[1,] -0.4  0.3  0.1 0   
[2,]  0.2 -0.5  0.2 0   
[3,]  0.2  0.2 -0.3 0

[,1] [,2] [,3] [,4]     
[1,] 1    0    0    0.3142857
[2,] 0    1    0    0.2857143
[3,] 0    0    1    0.4000000
[4,] 0    0    0    0.0000000

Matriz estacionaria
[1] 0.3142857 0.2857143 0.4000000


### **AUTOVALORES Y AUTOVECTORES**

In [ ]:
P=matrix(c(1/2,0,1/4,1/4,1,1/4,1/4,0,1/2),nrow=3,ncol=3)
P

[,1] [,2] [,3]
[1,] 0.50 0.25 0.25
[2,] 0.00 1.00 0.00
[3,] 0.25 0.25 0.50

In [ ]:
M=eigen(P)
M

eigen() decomposition
$values
[1] 1.00 0.75 0.25

$vectors
          [,1]      [,2]       [,3]
[1,] 0.5773503 0.7071068 -0.7071068
[2,] 0.5773503 0.0000000  0.0000000
[3,] 0.5773503 0.7071068  0.7071068


In [ ]:
d=M$values
d

[1] 1.00 0.75 0.25

In [ ]:
d=c(1,0.25,0.75)
d

[1] 1.00 0.25 0.75

In [ ]:
D=diag(d)
D

[,1] [,2] [,3]
[1,] 1    0.00 0.00
[2,] 0    0.25 0.00
[3,] 0    0.00 0.75

In [ ]:
C=M$vectors
C

[,1]      [,2]      [,3]      
[1,] 0.5773503 0.7071068 -0.7071068
[2,] 0.5773503 0.0000000  0.0000000
[3,] 0.5773503 0.7071068  0.7071068

In [ ]:
C[,2]

[1] 0.7071068 0.0000000 0.7071068

In [ ]:
C1=cbind(C[,1],C[,3],C[,2])
C1

[,1]      [,2]       [,3]     
[1,] 0.5773503 -0.7071068 0.7071068
[2,] 0.5773503  0.0000000 0.0000000
[3,] 0.5773503  0.7071068 0.7071068

In [ ]:
C1inv=solve(C1)
C1inv

[,1]       [,2]      [,3]     
[1,]  0.0000000  1.732051 0.0000000
[2,] -0.7071068  0.000000 0.7071068
[3,]  0.7071068 -1.414214 0.7071068

In [ ]:
(C1%*%D)%*%C1inv

[,1] [,2] [,3]
[1,] 0.50 0.25 0.25
[2,] 0.00 1.00 0.00
[3,] 0.25 0.25 0.50

In [ ]:
Dk=matrix(c(1,0,0,0,0,0,0,0,0),nrow=3,ncol=3)
Dk

[,1] [,2] [,3]
[1,] 1    0    0   
[2,] 0    0    0   
[3,] 0    0    0

In [ ]:
Pb=(C1%*%Dk)%*%C1inv
Pb

[,1] [,2] [,3]
[1,] 0    1    0   
[2,] 0    1    0   
[3,] 0    1    0